### DISCUSSION TITLE
Tips to generate test hex files

### DISCUSSION THREAD

**Discussion by mcuee** [avrdudes/avrdude#1380] [Ideas]
I have collected some srecord tips from @stefanrueger and @MCUdude here (as well as my learning). I will try to draft a Wiki page later.
https://github.com/mcuee/avrdude/wiki

@stefanrueger 
Please help to take a look and add some other tips you have. Thanks. We do not need to limit with srecord even though it is the most popular.

**Comment by mcuee** [comment]
**How to geneerate test hex file using srecord**
* https://github.com/avrdudes/avrdude/issues/967#issuecomment-1129380825

Not woring under Windows MSYS2 MinGW but should work under Linux
```
$ head -512c /dev/urandom | srec_cat - -binary -o - -intel > /tmp/rnd.eep
```

Working for all OS with bash shell.
```
$ yes The quick brown fox jumps over the lazy dog             | head -512c | srec_cat - -binary -o - -intel >/tmp/en.eep
$ yes Z jagt im komplett verwahrlosten Taxi quer durch Bayern | head -512c | srec_cat - -binary -o - -intel >/tmp/de.eep
```

**More srecord tricks**
* https://github.com/avrdudes/avrdude/issues/391#issuecomment-1220928466
* https://github.com/avrdudes/avrdude/issues/1166#issuecomment-1298881437
* https://github.com/avrdudes/avrdude/issues/1068#issuecomment-1279880853
* https://github.com/avrdudes/avrdude/issues/425

```
srec_cat -generate 0x0000 0x001F -repeat_data 0x00 -o zeros_31bytes.hex -Intel -obs=16
srec_cat -generate 0x0000 0x000F -repeat_data 0x55 -o hex55s_15bytes.hex -Intel -obs=16
yes "hello, world" | head -c256 | srec_cat - -bin -offset 0x200 -o hello_world_256B_offset_0x200.hex -Intel
srec_cat -generate 0x0000 0x8000 -repeat_data 0x00 -o zeros32kB.hex -Intel
```

* https://github.com/avrdudes/avrdude/issues/1073
```
srec_cat test_202B.hex -intel -crop 0x0000 0x7FFF -offset 0x0200 -o test_202B_offset_0x200.hex -intel

$ avrdude -pm32a/At | grep flash.page_size
.ptmm	ATmega32A	flash	page_size	128

$ srec_cat ATTOBASICV234_m32-16MHZ-uart_btldr -i -output - -c_array | tr \\n " " | cut -d, -f128
 0xFF
```

* https://github.com/avrdudes/avrdude/issues/474#issuecomment-1159592117
```
srec_cat entest.srec -offset 0x021000 -o entest_offset_0x21000.srec
srec_cat entest_offset_0x21000.srec  -o entest_offset_0x21000.hex -intel
srec_cat entest_offset_0x11000.hex -intel -o - -c-arra dopp -Output_Word | tail -9 | head -4
```

Some more:
```
srec_cat urboot_atmega328p_autobaud_lednop_ur_vbl.hex -Intel --offset -0x0B00

```

**Comment by stefanrueger** [comment]
@mcuee We have come across a number of hex files that exposed problems for programming. Would be great to keep them in a separate collection for testing releases. The one from #1508 could be named `eeprom-test-stk500-usbtiny.hex`. There are other ones: sketch ending in pages of 0xff, sketch with holes, sketch with more than 128k, sketch with holes and exceeding 128k, eeprom with pages of 0xff, eeprom with a page starting with two 0xff (the problem in #1508)

**Reply by mcuee** [reply]
Good idea. Probably we can use this discussion to upload the hex files for testing.

**Comment by mcuee** [comment]
Now we have the test hex files in git main. I will close this one. Further test files can be added as an enhancement.
https://github.com/avrdudes/avrdude/tree/main/tools/test_files
